Add `--filter enabled=true` for `docker plugin ls`

yongtang opened the pull request:
**- What I did**

This fix adds `--filter enabled=true` to `docker plugin ls`, as was specified in #28624.

**- How I did it**

The related API and docs has been updated.

**- How to verify it**

An integration test has been added.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![2903397](https://cloud.githubusercontent.com/assets/6932348/20458394/0dee4b8c-ae59-11e6-8c43-0121075c6ab0.jpg)


This fix fixes #28624.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

albers commented:
@yongtang Wow, you're fast!
Does this also support the inverse variant `enabled=false`, only listing those plugins that are disabled? If not, please add this as well.

yongtang commented:
Thanks @albers. The PR has been updated. Now `enabled=true` will only show `enabled` plugins and `enabled=false` will only show `disabled` plugins. In case no `enabled` filter is specified, all plugins will be shown.

I also take a look at the plugin type. It seems now plugin takes `interface types`. I am not sure the interface types could be directly translated into the normal plugin types (like `volume`, `network`, etc)
```
$ docker plugin inspect tiborvass/no-remove
...
            "Interface": {
                "Socket": "plugin.sock",
                "Types": [
                    "docker.volumedriver/1.0"
                ]
            },
...

```

albers commented:
I just verified the `enabled` filter. It works perfectly.
If the `type` filter requires more prior investigation, how about adding it in another PR?

albers commented:
@yongtang Are you planning to add the `type` filter in this PR or in another?

yongtang commented:
@albers The PR has been updated to allow filtering based on types.
Based on the docs (`docs/extend/config.md`) only two types are supported at the moment:
* docker.volumedriver/1.0
* docker.authz/1.0

Please take a look and let me know if there are any issues.

albers commented:
@yongtang I cannot verify the `type` filter because I get an error message:
```bash
$ docker plugin ls --filter type=docker.volumedriver/1.0
Error response from daemon: Invalid filter 'type'
$ docker version
Client:
 Version:      1.14.0-dev
 API version:  1.26
 Go version:   go1.7.3
 Git commit:   dceb38a
 Built:        Wed Nov 23 09:31:03 2016
 OS/Arch:      linux/amd64

Server:
 Version:             1.14.0-dev
 API version:         1.26
 Minimum API version: 1.12
 Go version:          go1.7.3
 Git commit:          dceb38a
 Built:               Wed Nov 23 09:31:03 2016
 OS/Arch:             linux/amd64
 Experimental:        false
```
`docker version` shows that I'm using the correct commit. Does it really work for you?

I'm also in doubt that the filter values are useful. As a user, I would like to get a list of e.g. all volume plugins. The value `docker.volumedriver/1.0` is very technical and also too specifc. What if there are different versions of volumedrivers? Would the user have to specify `-f type=docker.volumedriver/1.0` and `-f type=docker.volumedriver/2.0` and possibly more values? The user would have to know all interface versions.

`docker inspect --type` uses the values `network` and `volume`. I think it is better to map the interface names to these known values and not to introduce new names for these concepts. Perhaps mapping could use regular expressions to decrease maintenance efforts.
As there is `dockerd --authorization-plugin`, I would suggest the value `authorization` instead of `docker.authz/1.0`

@thaJeztah WDYT, how should the type filters be named?

yongtang committed:
Add `--filter enabled=true` for `docker plugin ls`

This fix adds `--filter enabled=true` to `docker plugin ls`,
as was specified in 28624.

The related API and docs has been updated.

An integration test has been added.

This fix fixes 28624.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add `capability` filter to `docker plugin ls`

This fix adds `--filter capability=[volumedriver|authz]` to `docker plugin ls`.

The related docs has been updated.

An integration test has been added.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@albers In the last PR, there was an issue  when I did the rebase and one line was missing. That was why in the last PR type was not working. Sorry about.

I updated the PR but there are some further changes.

The plugin in docker consists of two versions:
* Legacy Plugin
   Those are the plugins that was installed through daemon. For example, the auth, volume, or network plugin. Those plugins will implement a "Implements" API to specify which feature it will support. one plugin support multiple features are possible. For example, it is quite common for one plugin to support `IpamDriver` and `NetworkDriver` at the same time.
   The documentation has been quite missing or incomplete for legacy plugin. Though I believe below is the list of possible values of "Implements":
    - `GraphDriver`
    - `VolumeDriver`
    - `IpamDriver`
    - `NetworkDriver`
    - `authz`

* New Plugin (V2)
    The new plugin system types consists of three fields:
    - `Capability` (Currently either `volumedriver` or `authz`)
    - `Prefix` (Currently `docker`
    - `Version` (Currently it is `1.0`)
   Combined of the above fields is the type: `<Prefix>.<Capability>/<Version>`. That is where `docker.volumedriver/1.0` comes from.

When `docker plugin ls` or `docker plugin install` is invoked, only V2 plugins are impacted. 

The closest match between New Plugin and Legacy Plugin probably is the `Capability` in V2 and `Implements` in Legacy.

I updated the PR to add the filter of `--filter capability=volumedriver`/`--filter capability=authz` (replaced the `type` field in the previous one). It is a separate commit in this PR for ease of review.

Please take a look and see if `capability` makes sense. And let me know if there are any alternative suggestions.

cc @vdemeester @thaJeztah

yongtang commented:
Just realized that the output of `docker info` does not include V2 Plugin for volume drivers at the moment. Will open a separate PR about the Plugins in `docker info` shortly.

yongtang commented:
Added a PR #28764 to address the issue for the output of Plugin sections in `docker info`.

albers commented:
@yongtang Now it works. 
I only checked with the tiborvass/no-remove plugin. Do you know an auth plugin I could use for testing?

Thanks for the excellent elaboration on the plugin APIs. I like your suggestion of a `capability` filter.

yongtang commented:
Thanks @albers. I created a pass-through auth plugin `yongtang/no-authz` that could be used for test:
https://hub.docker.com/r/yongtang/no-authz/
https://github.com/yongtang/no-authz

```
ubuntu@ubuntu:~/$ docker plugin install yongtang/no-authz
Plugin "yongtang/no-authz" is requesting the following privileges:
 - network: [host]
Do you grant the above permissions? [y/N] y
yongtang/no-authz
ubuntu@ubuntu:~/$ docker plugin install tiborvass/no-remove
Plugin "tiborvass/no-remove" is requesting the following privileges:
 - network: []
 - mount: [/data]
Do you grant the above permissions? [y/N] y
tiborvass/no-remove
ubuntu@ubuntu:~/$ docker plugin ls
NAME                  TAG                 DESCRIPTION                  ENABLED
yongtang/no-authz     latest              No-AuthZ plugin for Docker   true
tiborvass/no-remove   latest              A test plugin for Docker     true
ubuntu@ubuntu:~/$ docker plugin ls --filter capability=volumedriver
NAME                  TAG                 DESCRIPTION                ENABLED
tiborvass/no-remove   latest              A test plugin for Docker   true
ubuntu@ubuntu:~/$ docker plugin ls --filter capability=authdriver
NAME                TAG                 DESCRIPTION                  ENABLED
yongtang/no-authz   latest              No-AuthZ plugin for Docker   true
ubuntu@ubuntu:~/$ 
```

albers commented:
@yongtang Wow, thanks a log for providing a sample authz plugin!
I verified the `--filter`s and everything works perfectly!
LGTM

vdemeester commented:
Desgin LGTM but I think this will be for 1.14.0 though :angel:

thaJeztah commented:
looks like this needs a rebase, @yongtang :cry:

yongtang commented:
Thanks @thaJeztah. The PR has been rebased.

vdemeester commented on plugin/backend_linux.go:
```diff
@@ -228,10 +234,45 @@ func (pm *Manager) Pull(name string, metaHeader http.Header, authConfig *types.A
 
 // List displays the list of plugins and associated metadata.
-func (pm *Manager) List() ([]types.Plugin, error) {
+func (pm *Manager) List(filter string) ([]types.Plugin, error) {
```
I think we should pass here a `filters.Args` instead of a string that is a json one. This is what we do on other filters (like on containers, …)

yongtang replied:
Thanks @vdemeester. The PR has been updated.

vdemeester commented on plugin/backend_linux.go:
```diff
@@ -228,10 +234,45 @@ func (pm *Manager) Pull(name string, metaHeader http.Header, authConfig *types.A
+	enabledOnly := false
+	disabledOnly := false
+	if pluginFilters.Include("enabled") {
+		if pluginFilters.ExactMatch("enabled", "true") || pluginFilters.ExactMatch("enabled", "1") {
```
I think we only allow `enable` on boolean filters (like `dangling`). I would expect the same here.

yongtang replied:
Thanks. Done.

vdemeester commented on client/plugin_list.go:
```diff
@@ -2,15 +2,26 @@ package client
 
 // PluginList returns the installed plugins
-func (cli *Client) PluginList(ctx context.Context) (types.PluginsListResponse, error) {
+func (cli *Client) PluginList(ctx context.Context, filter filters.Args) (types.PluginsListResponse, error) {
```
Could we add unit tests to validate the filters are sent (see other client method tests) 👼

yongtang replied:
Thanks. The unit test has been updated.

vdemeester reviewed (changes requested):
SGTM 👼
Needs a rebase and added a few comments

vdemeester commented on plugin/backend_linux.go:
```diff
@@ -259,10 +265,35 @@ func (pm *Manager) Pull(ctx context.Context, ref reference.Named, name string, m
+		return nil, err
+	}
+
+	enabledOnly := false
```
I think you misunderstood my comment :sweat_smile: I just think we don't support `1` and `0` for boolean filters (`--filter dangling=true` works, `--filter dangling=1` does not).

vdemeester reviewed (changes requested):
The code as of now will return all plugin if `--filter enable=false` ? (I would think it should only return plugin that are not enabled)

yongtang commented:
@vdemeester I was looking into the keyword `dangling` and was mislead by the following`dangling`: 😅 
https://github.com/docker/docker/blob/v1.13.0-rc4/daemon/list.go#L641-L649
```
	danglingOnly := false
	if filter.Include("dangling") {
		if filter.ExactMatch("dangling", "true") || filter.ExactMatch("dangling", "1") {
			danglingOnly = true
		} else if !filter.ExactMatch("dangling", "false") && !filter.ExactMatch("dangling", "0") {
			return nil, fmt.Errorf("Invalid filter 'dangling=%s'", filter.Get("dangling"))
		}
		retVols = daemon.volumes.FilterByUsed(retVols, !danglingOnly)
	}
```

The PR has been updated now. Please take a look.

vdemeester commented:
@yongtang oh I didn't know that.. Yeah I'm all for consistency here. Summoning @thaJeztah to get opinion 👼

vdemeester reviewed (approved):
LGTM 🦁

thaJeztah commented on docs/reference/commandline/plugin_ls.md:
```diff
@@ -40,6 +43,35 @@ ID                  NAME                  TAG                 DESCRIPTION
+### capability
+
+The `capability` filter matches on plugin capabilitiess. One plugin
+might have multiple capabilities. Currently `volumedriver` and `authz`
```
I think we support network plugins now as well?

thaJeztah replied:
Looks like it's not documented though; https://github.com/docker/docker/blob/master/docs/extend/config.md#config-field-descriptions

@vieux?

yongtang replied:
@thaJeztah @vieux I think both `network` driver and `ipam` driver are supported (though they are not documented). 

Also legacy plugin supports `graph` driver. Not sure if v2 plugin support graph driver or not.

cpuguy83 replied:
v2 plugins support graph driver, but only in experimental.

yongtang replied:
Thanks @cpuguy83 @thaJeztah. I updated the related docs so that `networkdriver` and `ipamdriver` are supported capabilities as well. (`graphdriver` was not added explicitly for its in experimental)

Also, `docs/extend/config.md` has been updated so that `networkdriver` and `ipamdriver` are supported plugin interfaces.

Please take a look.

thaJeztah commented on docs/reference/commandline/plugin_ls.md:
```diff
@@ -40,6 +43,35 @@ ID                  NAME                  TAG                 DESCRIPTION
+
+### capability
+
+The `capability` filter matches on plugin capabilitiess. One plugin
```
typo `capabilitiess`

yongtang replied:
Thanks @thaJeztah. The PR has been updated.

LK4D4 commented:
@yongtang need rebase

yongtang commented:
@LK4D4 The PR has been rebased. Thanks.

LK4D4 commented:
experimental failure is due hub problems

LK4D4 commented:
I don't know how docs stuff happens this days, so ping @thaJeztah

albers commented on docs/reference/commandline/plugin_ls.md:
````diff
@@ -42,6 +45,35 @@ ID                  NAME                             TAG                 DESCRIP
+`ipamdriver`, and `authz` are supported capabilities.
+
+```bash
+$ docker plugin install tiborvass/no-remove --disable
````
The options should go before the argument:
`docker plugin install --disable tiborvass/no-remove`

yongtang replied:
@albers The docs has been updated. Thanks.

vdemeester commented:
Moving to docs-review /cc @thaJeztah

vdemeester reviewed (approved):
docs LGTM 🐸
/cc @thaJeztah for docs revisit

thaJeztah commented:
Thanks @yongtang, docs LGTM, but could you also create a markdown for the man-pages, with just the "example" and "description", similar to https://github.com/docker/docker/blob/master/man/src/container/ls.md

That file will have to be in a "plugins" subdirectory inside man/src/
